near-merkle-proofs/src/lib.rs

```diff
@@ -0,0 +1,757 @@
+    }
+
+    /// Computes the root hash of a given merkle proof and item hash
+    /// It will update the cache of intermediate nodes so that they do not have
+    /// to be recomputed
+    pub fn calculate_root_hash(&mut self, proof: &MerklePath, item_hash: CryptoHash) -> CryptoHash {
+        // trivial example, where proof is empty
+        if proof.len() == 0 {
+            return CryptoHash::default();
+        }
+
+        // the first element is somewhat different, since the caller is passing the item's hash
+        let (_, node_coordinates_to_calculate) = self.get_node_coordinates(proof);
+        let nodes_to_calculate = node_coordinates_to_calculate.len();
+
+        let sibling_item = &proof[0];
+
+        // calculate the hash for the leaf level by hashing the item_hash given and its sibling (provided in the proof)
+        let hash = match sibling_item.direction {
+            Direction::Left => hash_borsh::<_, HF>(&(sibling_item.hash, item_hash)),
+            Direction::Right => hash_borsh::<_, HF>(&(item_hash, sibling_item.hash)),
+        };
+
+        let NodeCoordinates { index, level, .. } =
+            &node_coordinates_to_calculate[nodes_to_calculate - 0 - 1];
+        let cached_value = self.cached_nodes.inner.get(&(*level, *index));
+
+        match cached_value {
+            None => {
+                self.cached_nodes.inner.insert((*level, *index), hash);
+            }
+            Some(parent_hash) => {
+                // ensure that, if the value was cached it matches the calculation made above
+                // this is important, otherwise when most of the intermediates nodes are cached, if this check
+                // is not made, a wrong proof could be passed and stil "yield" the right root hash
+                assert_eq!(parent_hash, &hash);
+            }
+        }
+
+        proof
+            .iter()
+            .enumerate()
+            .skip(1) // skip the parent
+            .fold(hash, |mut hash, (item_idx, merkle_path_item)| {
+                let NodeCoordinates { index, level, .. } =
+                    &node_coordinates_to_calculate[nodes_to_calculate - item_idx - 1];
+
+                let cached_value = self.cached_nodes.inner.get(&(*level, *index));
+                match cached_value {
+                    None => {
+                        match merkle_path_item.direction {
+                            Direction::Left => {
+                                hash = CryptoHash::hash_borsh(&(merkle_path_item.hash, hash))
+                            }
+                            Direction::Right => {
+                                hash = CryptoHash::hash_borsh(&(hash, merkle_path_item.hash))
+                            }
+                        };
+                        // update the cache
+                        self.cached_nodes.inner.insert((*level, *index), hash);
+                    }
+                    Some(cached_value) => {
+                        hash = *cached_value;
```

Comment:
Shouldn't we validate the cached value (as on line 122)? Since we're not doing it in the `extend_from_given` step

Reply:
Thanks for your suggestion. I've added non happy path tests and included this